### ISSUE TITLE
Functionality Available?: Dataset Input perform slicing along time axis

### ISSUE THREAD

**Issue by Gumilton** [tensorflow/tensorflow#14906]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: Yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Windows 7
- **TensorFlow installed from (source or binary)**: from pip
- **TensorFlow version (use command below)**: 1.3 
- **Python version**: 3.6
- **Bazel version (if compiling from source)**:
- **GCC/Compiler version (if compiling from source)**:
- **CUDA/cuDNN version**: 8.0, 6.1
- **GPU model and memory**: k2200
- **Exact command to reproduce**:

### Describe the problem

Currently, as many suggested, when we want to train LSTM to predict the values of next time step, we would better slice all the samples along time axis to tuples of (lookback numbers of values, target predicted values) as (x, y), then store all these in a variable in RAM or as data file in disk. Then we build dataset to point to either form. However, this method makes LSTM stateless. Currently, our scheme to stateful LSTM is as follows:
for an input signal [1, 2, 3, 4, 5, 6, ..., 9, 10, 11]
we initialize LSTM and set state to 0.
feed ([1, 2, 3], 4) to train (`tf.nn.dynamic_rnn`), then feed ([2, 3, 4], 5) ....... ([8, 9, 10], 11), until end of this sequence.
In this batch, we have a number (Batch size, like 32) of signals like this and they will be processed in parallel in GPU by TF.
In the next batch of new signal samples, we reset LSTM with initial state being 0. And then repeat this process.

In this way, we think that given [8, 9, 10] to the model to predict next value (as 11), it is helpful for the model to choose whether to utilize the information of its current state (whether it is at beginning of a series of signal, zero initial; or it is at middle of a signal sequence).

Currently, before version 1.4, we built two generators, one (batch generator) is to generate a batch of signals. The other (time slicer) is to generate (x, y) [shape of (batch size, max time step, number of features) ] along time axis by using the data yield by the batch generator.

In version 1.4, we find Dataset, Estimator and Experiment pipeline powerful. Is there a way to implement the same idea using such pipeline?
Thanks!

**Comment by mrry** [comment]
Currently the `tf.data` APIs all treat the outermost dimension as the batch dimension, both for slicing into elements and for batching. Can you use `tf.tranpose()` or `np.transpose()` to reorder the dimensions so that this works for your data?

**Comment by Gumilton** [comment]
Thank you! @mrry 
Our data now is in shape [batch, time, features]. If we do a transpose into [time, batch, features], this is not what we want to have. 

Otherwise, for a batch of two signals of time series, for example, for simplicity number of features == 1, so of shape [2, 6, 1]
[ 
  [[1], [2], [3], [4], [5], [6]],
  [[7], [8], [9], [10], [11], [12]]
]
What we want to pass as input of x is
[
 [[1], [2], [3]],
 [[2], [3], [4]],
 [[3], [4], [5]],
 [[7], [8], [9]],
 [[8], [9], [10]],
 [[9], [10], [11]]
]
so of shape [batch * (max time - lookback), max time, num features], as [6, 6, 1]
How could we achieve this by using Dataset API? Thanks!

**Comment by mrry** [comment]
Ah, I misunderstood what you were asking for. Yes, you can do this using `Dataset.flat_map()`. For example, the following program produces the result you describe (although it doesn't use a `lookback` variable, so you might need to change it still):

```python
max_time = 3

def get_slices(x):
  num_slices = tf.shape(x, out_type=tf.int64)[0] - max_time
  return tf.data.Dataset.range(num_slices).map(lambda i: x[i:i + max_time])

dataset = tf.data.Dataset.from_tensor_slices([
    [[1], [2], [3], [4], [5], [6]],
    [[7], [8], [9], [10], [11], [12]]]).flat_map(get_slices).batch(6)
```

**Comment by Gumilton** [comment]
@mrry Thanks! This works in splitting them into chunks along time axis. However, if the lengths of the original signals are different, how could we know when to reset the state of LSTM to 0?
For example, if we have two batches and each batch size is 1. 
[
[[[1], [2], [3], [4], [5], [6]]],
[[[7], [8], [9], [10], [11], [12], [13]]]
]
The only one sample in the first batch is of length 6 while the other one is of length 7. We want reset the state of LSTM to 0 after slicing over the first sample, to make the LSTM state-ful.
